### PR TITLE
add power on after reset failure for rflash -d

### PR DESCRIPTION
**Pull request by bybai**
for #5299 
After rpower reset, check if 12 seconds later and power state is still off, send the rpower on request to let host on to resolve the problem.
1, positive UT:
```
[root@briggs01 baiyuan]# rflash mid05tor12cn13,mid05tor12cn16 -d fw/
Attempting to upload /root/baiyuan/fw/obmc-phosphor-image-witherspoon.ubi.mtd.tar and /root/baiyuan/fw/witherspoon-IBM-OP9_v1.19_1.172.pnor.squashfs.tar, please wait...
mid05tor12cn16: Firmware upload successful. Attempting to activate firmware: IBM-witherspoon-ibm-OP9_v1.19_1.172 (ID: d8368f51)
mid05tor12cn16: Firmware upload successful. Attempting to activate firmware: ibm-v2.0-0-r46-0-gbed584c (ID: b04fff27)
mid05tor12cn16: Firmware IBM-witherspoon-ibm-OP9_v1.19_1.172 activation successful.
mid05tor12cn16: Firmware ibm-v2.0-0-r46-0-gbed584c activation successful.
mid05tor12cn16: BMC reboot
mid05tor12cn13: Firmware upload successful. Attempting to activate firmware: IBM-witherspoon-ibm-OP9_v1.19_1.172 (ID: d8368f51)
mid05tor12cn13: Firmware upload successful. Attempting to activate firmware: ibm-v2.0-0-r46-0-gbed584c (ID: b04fff27)
mid05tor12cn13: Firmware IBM-witherspoon-ibm-OP9_v1.19_1.172 activation successful.
mid05tor12cn13: Firmware ibm-v2.0-0-r46-0-gbed584c activation successful.
mid05tor12cn13: BMC reboot
mid05tor12cn16: Retry BMC state, wait for 15 seconds ...
mid05tor12cn13: Retry BMC state, wait for 15 seconds ...
mid05tor12cn16: Retry BMC state, wait for 15 seconds ...
mid05tor12cn13: Retry BMC state, wait for 15 seconds ...
mid05tor12cn16: Retry BMC state, wait for 15 seconds ...
mid05tor12cn13: Retry BMC state, wait for 15 seconds ...
mid05tor12cn13: Retry BMC state, wait for 15 seconds ...
mid05tor12cn16: BMC Ready
mid05tor12cn16: reset
mid05tor12cn13: BMC Ready
mid05tor12cn13: reset
mid05tor12cn16: on
mid05tor12cn13: on
```
2, hack ``$all_status = $::POWER_STATE_OFF`` in code, and activate failure code pass, the UT like:
```
]# rflash mid05tor12cn13,mid05tor12cn16 -d fw/
Attempting to upload /root/baiyuan/fw/obmc-phosphor-image-witherspoon.ubi.mtd.tar and /root/baiyuan/fw/witherspoon-IBM-OP9_v1.19_1.172.pnor.squashfs.tar, please wait...
mid05tor12cn16: Firmware upload successful. Attempting to activate firmware: IBM-witherspoon-ibm-OP9_v1.19_1.172 (ID: d8368f51)
mid05tor12cn16: Firmware upload successful. Attempting to activate firmware: ibm-v2.0-0-r46-0-gbed584c (ID: b04fff27)
mid05tor12cn16: Firmware IBM-witherspoon-ibm-OP9_v1.19_1.172 activation successful.
mid05tor12cn16: Firmware ibm-v2.0-0-r46-0-gbed584c activation successful.
mid05tor12cn16: BMC reboot
mid05tor12cn13: Firmware upload successful. Attempting to activate firmware: IBM-witherspoon-ibm-OP9_v1.19_1.172 (ID: d8368f51)
mid05tor12cn13: Firmware upload successful. Attempting to activate firmware: ibm-v2.0-0-r46-0-gbed584c (ID: b04fff27)
mid05tor12cn13: Firmware IBM-witherspoon-ibm-OP9_v1.19_1.172 activation successful.
mid05tor12cn13: Firmware ibm-v2.0-0-r46-0-gbed584c activation successful.
mid05tor12cn13: BMC reboot
mid05tor12cn13: Retry BMC state, wait for 15 seconds ...
mid05tor12cn13: Retry BMC state, wait for 15 seconds ...
mid05tor12cn13: Retry BMC state, wait for 15 seconds ...
mid05tor12cn13: Retry BMC state, wait for 15 seconds ...
mid05tor12cn13: BMC Ready
mid05tor12cn13: reset
mid05tor12cn13: on
mid05tor12cn13: on
mid05tor12cn13: on
mid05tor12cn13: on
mid05tor12cn13: on
mid05tor12cn13: [briggs01]: Error: Sent power-on command but state did not change to on after waiting 70 seconds. (State=off).
```

### CODE REVIEW

File: xCAT-server/lib/xcat/plugins/openbmc.pm [hunk 1]
```diff
@@ -2639,7 +2657,7 @@ sub rpower_response {
     xCAT_monitoring::monitorctrl::setNodeStatusAttributes(\%new_status, 1) if (%new_status);
 
     my $all_status;
-    if ($node_info{$node}{cur_status} eq "RPOWER_STATUS_RESPONSE" or $node_info{$node}{cur_status} eq "RPOWER_CHECK_RESPONSE" or $node_info{$node}{cur_status} eq "RPOWER_BMC_STATUS_RESPONSE") {
+    if ($node_info{$node}{cur_status} eq "RPOWER_STATUS_RESPONSE" or $node_info{$node}{cur_status} eq "RPOWER_CHECK_RESPONSE" or $node_info{$node}{cur_status} eq "RPOWER_BMC_STATUS_RESPONSE" or $node_info{$node}{cur_status} eq "RPOWER_CHECK_ON_RESPONSE") {
```
Comment (gurevichmark) [inline xCAT-server/lib/xcat/plugins/openbmc.pm]: I think it would be useful to have some comments here about what is being done. This code is very hard to follow.

File: xCAT-server/lib/xcat/plugins/openbmc.pm [hunk 2]
```diff
@@ -2778,6 +2796,29 @@ sub rpower_response {
             } else {
                 $node_info{$node}{cur_status} = $next_status{ $node_info{$node}{cur_status} }{ON};
             }
+        } elsif ($node_info{$node}{cur_status} eq "RPOWER_CHECK_ON_RESPONSE") {
```
Comment (gurevichmark) [inline xCAT-server/lib/xcat/plugins/openbmc.pm]: I think it would be useful to have some comments here about what is being done. This code is very hard to follow.

File: xCAT-server/lib/xcat/plugins/openbmc.pm [hunk 3]
```diff
@@ -2778,6 +2796,29 @@ sub rpower_response {
...
+                    if ($node_info{$node}{wait_on_start}) {
+                        $node_info{$node}{wait_on_end} = time();
+                    } else {
+                        $node_info{$node}{wait_on_start} = time();
+                    }
+                    retry_after($node, $next_status{ $node_info{$node}{cur_status} }{OFF}, $::RPOWER_CHECK_ON_INTERVAL);
+                    return;
+                } else {
+                    my $wait_time_X = $node_info{$node}{wait_on_end} - $node_info{$node}{wait_on_start};
+                    xCAT::SvrUtils::sendmsg([1, "Sent power-on command but state did not change to $::POWER_STATE_ON after waiting $wait_time_X seconds. (State=$all_status)."], $callback, $node);
```
Comment (gurevichmark) [inline xCAT-server/lib/xcat/plugins/openbmc.pm]: I think we should add "BMC" to this message, so it is clear what is being powered on. Maybe "Sent power-on command to BMC, but state ..." and at the end "(BMC State=...)"

Comment (bybai) [reply xCAT-server/lib/xcat/plugins/openbmc.pm]: @gurevichmark , this `rpower on` is for host, it is not for `BMC`. After `rpower reset`, some times, host is off but it is not on. So here check if host is on after `rpower reset`, if it is off, then wait for 12 seconds, and execute `rpower on`. When host is `off, on`, the new PNOR can become active, or else it is `pending +` state.

**Comment by xuweibj** [comment]
Looks good to me. Thanks @bybai .
